### PR TITLE
CORTX-33479: F-20E:TA for IAM policy + Object versioning

### PR DESCRIPTION
**Pull request by AshwiniS-Borse**
Signed-off-by: AshwiniS-Borse <ashwini.borse@seagate.com>
TA: TA for IAM policy + Object versioning
[TEST-46690](https://jts.seagate.com/browse/TEST-46690) - Test Allow and Get Object Version API using IAM policy for self
[TEST-46689](https://jts.seagate.com/browse/TEST-46689) - Test Allow and Deny Get Object Version API using IAM policy for others
[TEST-46688](https://jts.seagate.com/browse/TEST-46688) - Test Allow and Deny Delete Object Version API using IAM policy for self
[TEST-46687](https://jts.seagate.com/browse/TEST-46687) - Test Allow and Deny Delete Object Version API using IAM policy for others.

TE will be covered in another ticket.

### CODE REVIEW

**Comment by AshwiniS-Borse** [comment]
@apoorva-rao @ravindrachoudhari @ketanarlulkar @kachhwahadivya please review PR

File: s3tests_boto3/functional/test_iam.py [hunk 1]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+        {"Version": "2012-10-17",
+         "Statement": {
+             "Effect": "Deny",
+             "Action": ["s3:GetObjectVersion"],
+             "Resource": f"arn:aws:s3:::{bucket}"
+         }
+         }
+    )
+    response = client.put_user_policy(PolicyDocument=deny_get_object_versioning_policy,
+                                      PolicyName='deny_policy_versioning',
```
Comment (ketanarlulkar) [inline s3tests_boto3/functional/test_iam.py:2084]: Policy name can not have '_' (underscore) in it. Ref: [AWS - IAM - PutUserPolicy](https://docs.aws.amazon.com/IAM/latest/APIReference/API_PutUserPolicy.html)

File: s3tests_boto3/functional/test_iam.py [hunk 2]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+def test_allow_deny_get_object_version_iam_policy_self():
+    client = get_iam_client()
+    s3_client_iam = get_iam_s3client()
+    bucket = get_new_bucket(client=s3_client_iam)
+    response = s3_client_iam.put_bucket_versioning(Bucket=bucket,
+                                                   VersioningConfiguration={"Status": "Enabled"})
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    s3_client_iam.put_object(Bucket=bucket, Key='iam1buk1obj1', Body='bar')
+    response = s3_client_iam.list_object_versions(Bucket=bucket, Prefix='iam1buk1obj1')
+    version_id = response['Versions'][0]['VersionId']
```
Comment (ketanarlulkar) [inline s3tests_boto3/functional/test_iam.py:2071]:
```suggestion
    version_id = response['VersionId']
```

File: s3tests_boto3/functional/test_iam.py (hunk 1 above)
Comment (AshwiniS-Borse) [reply s3tests_boto3/functional/test_iam.py:2084]: image

Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:2084]: My bad. Pattern just below has \w (which includes `_`)
> Pattern: [\w+=,.@-]+

**Review by ravindrachoudhari — APPROVED** [review]

**Review by ketanarlulkar — APPROVED** [review]

File: s3tests_boto3/functional/test_iam.py [hunk 3]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+    status, error_code = _get_status_and_error_code(e.response)
+    eq(status, 403)
+    eq(error_code, 'AccessDenied')
+    response = s3_client_iam.head_object(Bucket=bucket, Key='iam1buk1obj1')
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='deny_policy_versioning',
+                                         UserName=get_iam_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+    response = s3_client_iam.get_object(Bucket=bucket, Key='iam1buk1obj1')
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: Let's move GET Object verification at lines 2103 - 2104 after line 2098 and before we delete policy (at 2099)

File: s3tests_boto3/functional/test_iam.py [hunk 4]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+                                                   VersioningConfiguration={"Status": "Enabled"})
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    s3_client_iam.put_object(Bucket=bucket, Key='iam1buk1obj1', Body='bar')
+    s3_client_iam.put_object(Bucket=bucket, Key='iam1buk1obj1', Body='bar')
+    response = s3_client_iam.list_object_versions(Bucket=bucket, Prefix='iam1buk1obj1')
+    version_id_1 = response['Versions'][0]['VersionId']
+    version_id_2 = response['Versions'][1]['VersionId']
+    response = s3_client_iam.delete_object(Bucket=bucket, Key='iam1buk1obj1',
+                                           VersionId=version_id_1)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: Status code here for successful deletion would be 204

File: s3tests_boto3/functional/test_iam.py [hunk 5]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+                                      PolicyName='allow_policy_versioning',
+                                      UserName=get_iam_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    e = assert_raises(ClientError, s3_client_iam.delete_object, Bucket=bucket,
+                      Key='iam1buk1obj1', VersionId=version_id_1)
+    status, error_code = _get_status_and_error_code(e.response)
+    eq(error_code, 'NoSuchKey')
+    response = s3_client_iam.delete_object(Bucket=bucket, Key='iam1buk1obj1',
+                                           VersionId=version_id_2)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: Status code needs to be corrected to 204

File: s3tests_boto3/functional/test_iam.py [hunk 6]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+                                                   VersioningConfiguration={"Status": "Enabled"})
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    s3_client_alt.put_object(Bucket=bucket, Key='iam2buk1obj1', Body='foobar')
+    s3_client_alt.put_object(Bucket=bucket, Key='iam2buk1obj1', Body='foobar')
+    response = s3_client_alt.list_object_versions(Bucket=bucket, Prefix='iam2buk1obj1')
+    version_id_1 = response['Versions'][0]['VersionId']
+    version_id_2 = response['Versions'][1]['VersionId']
+    response = s3_client_alt.delete_object(Bucket=bucket, Key='iam2buk1obj1',
+                                           VersionId=version_id_1)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: Status code would be 204

File: s3tests_boto3/functional/test_iam.py [hunk 7]
```diff
@@ -2052,3 +2052,290 @@ def test_allow_deny_put_bucket_versioning_iam_policy_others():
...
+         }
+    )
+    response = client.put_user_policy(PolicyDocument=allow_delete_object_versioning_policy,
+                                      PolicyName='policy_put_versioning',
+                                      UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_alt.get_object(Bucket=bucket, Key='iam2buk1obj1',
+                                        VersionId=version_id_2)
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: We need to test DELETE Object with version ID specified would succeed here for s3_client_alt.

**Review by apoorva-rao — CHANGES_REQUESTED** [review]
@AshwiniS-Borse please check the inline review comments

File: s3tests_boto3/functional/test_iam.py (hunk 3 above)
Comment (AshwiniS-Borse) [reply s3tests_boto3/functional/test_iam.py]: This is as per step 6 in confluence Page.https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/980255029/F-20E+S3+API+IAM+Policy+Support+-+PI7
get_object is already called at line number 2087

File: s3tests_boto3/functional/test_iam.py (hunk 4 above)
Comment (AshwiniS-Borse) [reply s3tests_boto3/functional/test_iam.py]: 200 status code is mentioned in confluence page https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/980255029/F-20E+S3+API+IAM+Policy+Support+-+PI7.
I have not ran these tests

File: s3tests_boto3/functional/test_iam.py (hunk 3 above)
Comment (apoorva-rao) [reply s3tests_boto3/functional/test_iam.py]: At line 2087 we are calling GET Object with version id. At line 2103 we were verifying GET Object without version id, which needs to be tested when we deny GetObjectVersion. I see the changes have been  made, so no further action needed here.

File: s3tests_boto3/functional/test_iam.py (hunk 4 above)
Comment (apoorva-rao) [reply s3tests_boto3/functional/test_iam.py]: Added a comment to TDS that status code for DELETE Object needs to be refactored to 204 No Content. I see the code has been updated here, so we're good.

**Review by apoorva-rao — APPROVED** [review]
LGTM

**Pull request merged**